### PR TITLE
Shoryuken adapter support

### PR DESCRIPTION
**Pull request by carlosantoniodasilva**
Integrates Judoscale with Shoryuken using Amazon SQS, by reporting queue depth. 

There's a few things to be aware with shoryuken currently:

- There doesn't seem to be an easy way to fetch or calculate a queue latency, so we're starting with queue depth and will investigate potential options for queue latency in the future.
- Tracking busy jobs doesn't seem to be possible currently either, without some hacky ideas, so we're explicitly disabling that support and not tracking busy jobs with Shoryuken initially.
- Shoryuken loads its list of queues via the CLI or configured file, however that only works for the shoryuken process itself, other process (like a Rails web process) won't know about those queues, and thus won't report any queues. That means downscaling to 0 will prevent us from collecting Shoryuken metrics. The simpler solution is to explicitly list the queues for Judoscale to collect, across the board, which would then allow Rails web process as well as any other process to collect them as expected. We'll be documenting this gotcha, and possibly revisit this in the future.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -141,6 +146,7 @@ Judoscale.configure do |config|
   # Enables reporting for active (busy) workers so that downscaling can be
   # suppressed.
   # See https://judoscale.com/docs/long-running-jobs/.
+  # Note: judoscale-shoryuken does not support reporting busy jobs.
```
Comment (carlosantoniodasilva) [inline README.md:149]: I added a couple of notes here about the support / functionality.

File: README.md [hunk 2]
```diff
@@ -54,15 +56,16 @@ The middleware will start the async reporter when it processes the first request
...
 
 ```ruby
 gem "judoscale-sidekiq"
+gem "judoscale-solid_queue"
 gem "judoscale-resque"
 gem "judoscale-delayed_job"
 gem "judoscale-good_job"
 gem "judoscale-que"
-gem "judoscale-solid_queue"
+gem "judoscale-shoryuken"
```
Comment (carlosantoniodasilva) [inline README.md:68]: Reorded to match how we were listing them elsewhere.

File: judoscale-ruby/lib/judoscale/config.rb [hunk 3]
```diff
@@ -34,7 +36,15 @@ def reset
 
         # Support for deprecated legacy env var configs.
         @max_queues = (ENV["RAILS_AUTOSCALE_MAX_QUEUES"] || 20).to_i
-        @track_busy_jobs = ENV["RAILS_AUTOSCALE_LONG_JOBS"] == "true"
+        self.track_busy_jobs = ENV["RAILS_AUTOSCALE_LONG_JOBS"] == "true"
+      end
+
+      def track_busy_jobs=(value)
+        if value && !@support_busy_jobs
```
Comment (carlosantoniodasilva) [inline judoscale-ruby/lib/judoscale/config.rb:43]: Trying to set this to any truth value will raise for shoryuken.. I checked the other libs and all support busy jobs -- maybe we can have a follow-up to look into it some more in the future, if there's another way we can provide this bit of functionality?

File: judoscale-ruby/lib/judoscale/config.rb [hunk 4]
```diff
@@ -34,7 +36,15 @@ def reset
...
+          raise ArgumentError, "#{identifier} does not support busy jobs"
+        end
+
+        @track_busy_jobs = value
       end
 
       def as_json
```
Comment (carlosantoniodasilva) [inline judoscale-ruby/lib/judoscale/config.rb:50]: Should we exclude `track_busy_jobs` if unsupported?

File: sample-apps/delayed_job-sample/config/initializers/judoscale.rb
Comment (carlosantoniodasilva) [inline sample-apps/delayed_job-sample/config/initializers/judoscale.rb:1]: Rename these initializers.

File: sample-apps/shoryuken-sample/README.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+- [Heroku CLI](https://devcenter.heroku.com/articles/heroku-cli)
+- AWS Credentials / Region
+
+## Set up the app
+
+Run `./bin/setup` install necessary dependencies. This will...
+
+- Run `bundle install` to install gems
+
+### Shoryuken / SQS setup
```
Comment (carlosantoniodasilva) [inline sample-apps/shoryuken-sample/README.md:18]: Added some notes about AWS/SQS setup, let me know if there's anything missing or confusing.

File: judoscale-ruby/lib/judoscale/config.rb (hunk 3 above)
Comment (adamlogic) [reply judoscale-ruby/lib/judoscale/config.rb:43]: I think we might revisit if someone explicitly asks for it, but for now I don't think it's worth spending time on it.

File: judoscale-ruby/lib/judoscale/config.rb (hunk 4 above)
Comment (adamlogic) [reply judoscale-ruby/lib/judoscale/config.rb:50]: I don't think it matters either way, so I lean toward keeping it as-is.

File: judoscale-shoryuken/rails-autoscale-shoryuken.gemspec [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+$LOAD_PATH.unshift(lib) unless $LOAD_PATH.include?(lib)
+require "judoscale/shoryuken/version"
+
+Gem::Specification.new do |spec|
+  spec.name = "rails-autoscale-shoryuken"
+  spec.version = Judoscale::Shoryuken::VERSION
+  spec.authors = ["Adam McCrea", "Carlos Antonio da Silva", "Jon Sullivan"]
+  spec.email = ["hello@judoscale.com"]
+
+  spec.summary = "This gem provides Shoryuken integration with the Judoscale autoscaling add-on for Heroku."
```
Comment (adamlogic) [inline judoscale-shoryuken/rails-autoscale-shoryuken.gemspec:11]: We're no longer just "for Heroku", but I know this is the same summary we have for the other gems. I think I want to change all of their summaries to something short and sweet like "Autoscaling for [job backend] workers". Judoscale is already in the gem name, so I don't think we even need to mention it in the summary.

Do you mind sneaking that change in here or creating a separate PR for it?

File: sample-apps/shoryuken-sample/README.md (hunk 5 above)
Comment (adamlogic) [reply sample-apps/shoryuken-sample/README.md:18]: This is great!

File: sample-apps/shoryuken-sample/README.md [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+heroku create
+git push heroku main
+```
+
+To install Judoscale:
+
+```sh
+# scale up a worker dyno before doing this so Judoscale picks it up
+heroku ps:scale heroku_shoryuken=1
+heroku addons:create judoscale:trial
```
Comment (adamlogic) [inline sample-apps/shoryuken-sample/README.md]: I'm sure this was copied from our other sample apps, but we haven't had a `trial` plan in a long time. 😅 

```suggestion
heroku addons:create judoscale
```

Comment (adamlogic) [reply sample-apps/shoryuken-sample/README.md]: I also don't think anyone except us ever uses the sample apps, let alone deploys them to Heroku, so it really doesn't matter.

**Review by adamlogic — APPROVED** [review]
Fantastic work!! 🙌

File: sample-apps/shoryuken-sample/README.md (hunk 7 above)
Comment (carlosantoniodasilva) [reply sample-apps/shoryuken-sample/README.md]: Right! There were 3 mentions of `:trial` that I removed in a follow-up commit. 👍

File: judoscale-shoryuken/rails-autoscale-shoryuken.gemspec (hunk 6 above)
Comment (carlosantoniodasilva) [reply judoscale-shoryuken/rails-autoscale-shoryuken.gemspec:11]: Pushed separately just to make it easier to review: https://github.com/judoscale/judoscale-ruby/pull/206

**Comment by ClaytonPassmore** [comment]
This is great! We wrote a custom metric collector for this. Awesome to see it directly supported now 🙌

**Comment by adamlogic** [comment]
@ClaytonPassmore Thanks! Let us know when you get a chance to try the official gem and how it works for you.

**Pull request merged**